RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc

```diff
@@ -0,0 +1,226 @@
+  ~SiPixelRecHitSoAFromLegacy() override = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+   using HitModuleStart = std::array<uint32_t,gpuClustering::MaxNumModules + 1>;
+   using HMSstorage = HostProduct<unsigned int[]>;
+
+
+private:
+  void produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const override;
+
+  // The mess with inputs will be cleaned up when migrating to the new framework
+  edm::EDGetTokenT<reco::BeamSpot> bsGetToken_;
+  edm::EDGetTokenT<SiPixelClusterCollectionNew> clusterToken_;    // Legacy Clusters
+  edm::EDPutTokenT<TrackingRecHit2DHost> tokenHit_;
+  edm::EDPutTokenT<HMSstorage> tokenModuleStart_;
+
+  std::string cpeName_;
+
+};
+
+SiPixelRecHitSoAFromLegacy::SiPixelRecHitSoAFromLegacy(const edm::ParameterSet& iConfig)
+    : bsGetToken_{consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("beamSpot"))},
+      clusterToken_{consumes<SiPixelClusterCollectionNew>(iConfig.getParameter<edm::InputTag>("src"))},
+      tokenHit_{produces<TrackingRecHit2DHost>()},
+      tokenModuleStart_{produces<HMSstorage>()},
+      cpeName_(iConfig.getParameter<std::string>("CPE")) {}
+
+void SiPixelRecHitSoAFromLegacy::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("beamSpot", edm::InputTag("offlineBeamSpot"));
+  desc.add<edm::InputTag>("src", edm::InputTag("siPixelClustersPreSplitting"));
+  desc.add<std::string>("CPE", "PixelCPEFast");
+  descriptions.add("siPixelRecHitHostSoA", desc);
+}
+
+void SiPixelRecHitSoAFromLegacy::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& es) const {
+
+  
+  const TrackerGeometry *geom_ = nullptr;
+  const PixelClusterParameterEstimator* cpe_ = nullptr;
+
+  
+  edm::ESHandle<TrackerGeometry> geom;
+  es.get<TrackerDigiGeometryRecord>().get( geom );
+  geom_ = geom.product();
+  
+
+  edm::ESHandle<PixelClusterParameterEstimator> hCPE;
+  es.get<TkPixelCPERecord>().get(cpeName_, hCPE);
+  cpe_ = dynamic_cast<const PixelCPEBase*>(hCPE.product());
+
+  PixelCPEFast const* fcpe = dynamic_cast<const PixelCPEFast*>(cpe_);
+  if (!fcpe) {
+    throw cms::Exception("Configuration") << "too bad, not a fast cpe gpu processing not possible....";
+  }
+  auto cpeView = fcpe->getCPUProduct();
+
+  const reco::BeamSpot& bs = iEvent.get(bsGetToken_);
+
+
+  BeamSpotCUDA::Data bsHost;
+  bsHost.x = bs.x0();
+  bsHost.y = bs.y0();
+  bsHost.z = bs.z0();
+
+  auto const& input = iEvent.get(clusterToken_);
+
+  // yes a unique ptr of a unique ptr so edm is happy and the pointer stay still...
+  auto hmsp = std::make_unique<uint32_t[]>(gpuClustering::MaxNumModules + 1);
+  auto hitsModuleStart = hmsp.get();
+  auto hms = std::make_unique<HMSstorage>(std::move(hmsp)); // hmsp is gone
+  iEvent.put(tokenModuleStart_,std::move(hms));  // hms is gone! hitsModuleStart still alive and kicking...
```

Comment:
I think it would be more clear to move the two lines above (`auto hms ...` and `iEvent.put()` (or replacing with `iEvent.emplace()`)) to the end of produce(). Then one wouldn't have to care about the lifetimes of the unique_ptr-pointed-to data wrt. moves of the unique_ptrs.

Reply:
I started to like this pattern of keeping a C-ptr (or even a reference actually) and putting in the event as early as possible
1) is robust against early return
2) is close to its creation 
3) avoids to percolate a unique_ptr that exists only because of edm
Actually would be even more user friendly if ```iEvent.emplace()``` would return a reference to be used through-out the algorithm